File: crates/kuben-store/src/repo/throttle.rs
//! Login-throttle windows (scenario 1), shared by every replica. Buckets are
//! opaque keys chosen by the caller, which hashes them: no email address or
//! client IP is stored.

use crate::{Store, StoreError};

/// A fixed window: `failures` counted since `started_at` (unix ms).
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct ThrottleWindow {
    pub failures: i64,
    pub started_at: i64,
}

const SELECT_WINDOW: &str = "SELECT failures, started_at FROM login_throttle WHERE bucket = $1";
/// Count one failure atomically. A window that began at or before `$3`
/// (`now - window`) has expired and restarts at 1.
const RECORD_FAILURE: &str = "INSERT INTO login_throttle (bucket, failures, started_at) VALUES ($1, 1, $2) \
     ON CONFLICT (bucket) DO UPDATE SET \
     failures = CASE WHEN login_throttle.started_at > $3 THEN login_throttle.failures + 1 ELSE 1 END, \
     started_at = CASE WHEN login_throttle.started_at > $3 THEN login_throttle.started_at ELSE $2 END";
const CLEAR: &str = "DELETE FROM login_throttle WHERE bucket = $1";
const PURGE: &str = "DELETE FROM login_throttle WHERE started_at <= $1";

impl Store {
    pub async fn throttle_window(&self, bucket: &str) -> Result<Option<ThrottleWindow>, StoreError> {
        let row: Option<(i64, i64)> = sqlx::query_as(SELECT_WINDOW)
            .bind(bucket)
            .fetch_optional(self.pool())
            .await?;
        Ok(row.map(|(failures, started_at)| ThrottleWindow { failures, started_at }))
    }

    /// Record a failed attempt at `now`; windows that started at or before
    /// `window_start` are replaced by a new one.
    pub async fn throttle_record_failure(
        &self,
        bucket: &str,
        now: i64,
        window_start: i64,
    ) -> Result<(), StoreError> {
        sqlx::query(RECORD_FAILURE)
            .bind(bucket)
            .bind(now)
            .bind(window_start)
            .execute(self.pool())
            .await?;
        Ok(())
    }

    pub async fn throttle_clear(&self, bucket: &str) -> Result<(), StoreError> {
        sqlx::query(CLEAR).bind(bucket).execute(self.pool()).await?;
        Ok(())
    }

    /// Drop windows that started at or before `window_start` (expired).
    pub async fn throttle_purge(&self, window_start: i64) -> Result<u64, StoreError> {
        let done = sqlx::query(PURGE)
            .bind(window_start)
            .execute(self.pool())
            .await?
            .rows_affected();
        Ok(done)
    }
}

File: go/hub/internal/store/throttle.go
package store

// Login-throttle windows (scenario 1), shared by every replica; the port of
// repo/throttle.rs. Buckets are opaque keys chosen by the caller, which
// hashes them: no email address or client IP is stored.

import (
	"context"

	"github.com/jackc/pgx/v5"
)

// ThrottleWindow is a fixed window: Failures counted since StartedAt (unix
// ms).
type ThrottleWindow struct {
	Failures  int64
	StartedAt int64
}

const (
	selectWindow = "SELECT failures, started_at FROM login_throttle WHERE bucket = $1"
	// recordFailure counts one failure atomically. A window that began at or
	// before $3 (now - window) has expired and restarts at 1.
	recordFailure = "INSERT INTO login_throttle (bucket, failures, started_at) VALUES ($1, 1, $2) " +
		"ON CONFLICT (bucket) DO UPDATE SET " +
		"failures = CASE WHEN login_throttle.started_at > $3 THEN login_throttle.failures + 1 ELSE 1 END, " +
		"started_at = CASE WHEN login_throttle.started_at > $3 THEN login_throttle.started_at ELSE $2 END"
	clearWindow  = "DELETE FROM login_throttle WHERE bucket = $1"
	purgeWindows = "DELETE FROM login_throttle WHERE started_at <= $1"
)

// ThrottleWindow is bucket's current window, if it has one.
func (s *Store) ThrottleWindow(ctx context.Context, bucket string) (ThrottleWindow, bool, error) {
	return queryOpt(ctx, s.db, "read a throttle window", selectWindow,
		func(row pgx.CollectableRow) (ThrottleWindow, error) {
			var w ThrottleWindow
			err := row.Scan(&w.Failures, &w.StartedAt)
			return w, err
		}, bucket)
}

// ThrottleRecordFailure records a failed attempt at now; windows that
// started at or before windowStart are replaced by a new one.
func (s *Store) ThrottleRecordFailure(ctx context.Context, bucket string, now, windowStart int64) error {
	_, err := exec(ctx, s.db, "record a failed attempt", recordFailure, bucket, now, windowStart)
	return err
}

// ThrottleClear forgets bucket's window.
func (s *Store) ThrottleClear(ctx context.Context, bucket string) error {
	_, err := exec(ctx, s.db, "clear a throttle window", clearWindow, bucket)
	return err
}

// ThrottlePurge drops windows that started at or before windowStart
// (expired) and returns how many.
func (s *Store) ThrottlePurge(ctx context.Context, windowStart int64) (uint64, error) {
	return exec(ctx, s.db, "purge throttle windows", purgeWindows, windowStart)
}
